File: sdk/dart/lib/src/cloud_instance.dart
import 'http_client.dart';

/// Cloud Instance: full VM lifecycle management
class CloudInstanceResource {
  final HttpClient httpClient;
  CloudInstanceResource(this.httpClient);

  /// GET /api/service/{id}/vms - List virtual servers
  Future<Map<String, dynamic>> listVms([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms', params);
  }

  /// POST /api/service/{id}/vms - Create new virtual server
  Future<Map<String, dynamic>> createVm([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms', params);
  }

  /// GET /api/service/{id}/vms/{vmid} - Get VM details
  Future<Map<String, dynamic>> getVm([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}', params);
  }

  /// DELETE /api/service/{id}/vms/{vmid} - Remove virtual server
  Future<Map<String, dynamic>> destroyVm([Map<String, dynamic>? params]) async {
    return httpClient.delete('/api/service/{id}/vms/{vmid}', params);
  }

  /// PUT /api/service/{id}/vms/{vmid} - Edit/resize virtual server
  Future<Map<String, dynamic>> resizeVm([Map<String, dynamic>? params]) async {
    return httpClient.put('/api/service/{id}/vms/{vmid}', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/stop - Stop virtual server
  Future<Map<String, dynamic>> stopVm([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/stop', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/start - Start virtual server
  Future<Map<String, dynamic>> startVm([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/start', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/reboot - Reboot virtual server
  Future<Map<String, dynamic>> rebootVm([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/reboot', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/resetpwd - Reset root password
  Future<Map<String, dynamic>> resetPassword([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/resetpwd', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/rebuild - Rebuild server
  Future<Map<String, dynamic>> rebuildVm([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/rebuild', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/ips - List IPs assigned to VM
  Future<Map<String, dynamic>> listVmIps([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/ips', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/ips - Assign IP to VM
  Future<Map<String, dynamic>> assignIp([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/ips', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/interfaces - List network interfaces
  Future<Map<String, dynamic>> listInterfaces([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/interfaces', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/usage/cpu - CPU usage graph
  Future<Map<String, dynamic>> getCpuUsage([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/usage/cpu', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/usage/net - Network usage graph
  Future<Map<String, dynamic>> getNetworkUsage([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/usage/net', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/usage/disk - Disk usage graph
  Future<Map<String, dynamic>> getDiskUsage([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/usage/disk', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/storage - List disks attached to VM
  Future<Map<String, dynamic>> listDisks([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/storage', params);
  }

  /// PUT /api/service/{id}/vms/{vmid}/storage/{diskid} - Resize disk
  Future<Map<String, dynamic>> resizeDisk([Map<String, dynamic>? params]) async {
    return httpClient.put('/api/service/{id}/vms/{vmid}/storage/{diskid}', params);
  }

  /// GET /api/service/{id}/images - List ISO images
  Future<Map<String, dynamic>> listIsoImages([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/images', params);
  }

  /// POST /api/service/{id}/images - Add ISO image
  Future<Map<String, dynamic>> addIsoImage([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/images', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/interfaces/{iface}/ips - List available IPs
  Future<Map<String, dynamic>> listAvailableIps([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/interfaces/{iface}/ips', params);
  }

  /// DELETE /api/service/{id}/vms/{vmid}/ips/{ipid} - Remove IP from VM
  Future<Map<String, dynamic>> removeIp([Map<String, dynamic>? params]) async {
    return httpClient.delete('/api/service/{id}/vms/{vmid}/ips/{ipid}', params);
  }

  /// GET /api/service/{id}/vms/{vmid}/rdns - Get reverse DNS entries
  Future<Map<String, dynamic>> getRdns([Map<String, dynamic>? params]) async {
    return httpClient.get('/api/service/{id}/vms/{vmid}/rdns', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/rdns - Update reverse DNS entries
  Future<Map<String, dynamic>> updateRdns([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/rdns', params);
  }

  /// POST /api/service/{id}/vms/{vmid}/rebuild_network - Rebuild network
  Future<Map<String, dynamic>> rebuildNetwork([Map<String, dynamic>? params]) async {
    return httpClient.post('/api/service/{id}/vms/{vmid}/rebuild_network', params);
  }

}

File: sdk/rust/src/cloud_instance.rs
//! Cloud Instance: full VM lifecycle management

use std::sync::Arc;
use crate::http_client::HttpClient;

pub struct CloudInstanceResource {
    http: Arc<HttpClient>,
}

impl CloudInstanceResource {
    pub fn new(http: Arc<HttpClient>) -> Self { Self { http } }

    /// GET /api/service/{id}/vms - List virtual servers
    pub async fn list_vms(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms", None).await
    }

    /// POST /api/service/{id}/vms - Create new virtual server
    pub async fn create_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms", None).await
    }

    /// GET /api/service/{id}/vms/{vmid} - Get VM details
    pub async fn get_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}", None).await
    }

    /// DELETE /api/service/{id}/vms/{vmid} - Remove virtual server
    pub async fn destroy_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.delete("/api/service/{id}/vms/{vmid}", None).await
    }

    /// PUT /api/service/{id}/vms/{vmid} - Edit/resize virtual server
    pub async fn resize_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.put("/api/service/{id}/vms/{vmid}", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/stop - Stop virtual server
    pub async fn stop_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/stop", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/start - Start virtual server
    pub async fn start_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/start", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/reboot - Reboot virtual server
    pub async fn reboot_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/reboot", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/resetpwd - Reset root password
    pub async fn reset_password(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/resetpwd", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/rebuild - Rebuild server
    pub async fn rebuild_vm(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/rebuild", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/ips - List IPs assigned to VM
    pub async fn list_vm_ips(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/ips", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/ips - Assign IP to VM
    pub async fn assign_ip(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/ips", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/interfaces - List network interfaces
    pub async fn list_interfaces(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/interfaces", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/usage/cpu - CPU usage graph
    pub async fn get_cpu_usage(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/usage/cpu", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/usage/net - Network usage graph
    pub async fn get_network_usage(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/usage/net", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/usage/disk - Disk usage graph
    pub async fn get_disk_usage(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/usage/disk", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/storage - List disks attached to VM
    pub async fn list_disks(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/storage", None).await
    }

    /// PUT /api/service/{id}/vms/{vmid}/storage/{diskid} - Resize disk
    pub async fn resize_disk(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.put("/api/service/{id}/vms/{vmid}/storage/{diskid}", None).await
    }

    /// GET /api/service/{id}/images - List ISO images
    pub async fn list_iso_images(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/images", None).await
    }

    /// POST /api/service/{id}/images - Add ISO image
    pub async fn add_iso_image(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/images", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/interfaces/{iface}/ips - List available IPs
    pub async fn list_available_ips(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/interfaces/{iface}/ips", None).await
    }

    /// DELETE /api/service/{id}/vms/{vmid}/ips/{ipid} - Remove IP from VM
    pub async fn remove_ip(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.delete("/api/service/{id}/vms/{vmid}/ips/{ipid}", None).await
    }

    /// GET /api/service/{id}/vms/{vmid}/rdns - Get reverse DNS entries
    pub async fn get_rdns(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.get("/api/service/{id}/vms/{vmid}/rdns", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/rdns - Update reverse DNS entries
    pub async fn update_rdns(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/rdns", None).await
    }

    /// POST /api/service/{id}/vms/{vmid}/rebuild_network - Rebuild network
    pub async fn rebuild_network(&self, params: Option<&serde_json::Value>) -> Result<serde_json::Value, crate::http_client::ApiError> {
        self.http.post("/api/service/{id}/vms/{vmid}/rebuild_network", None).await
    }

}
